.github/workflows/pull-request.yaml

```diff
@@ -26,20 +26,47 @@ jobs:
       CARGO_PROFILE_DEV_DEBUG: 0
       CARGO_PROFILE_TEST_DEBUG: 0
       FLYWAY_VERSION: 7.3.1
+      FLYWAY_PATH: /home/runner/flyway
       # Flyway is connecting via TCP/IP (authentication via password), whereas psql connects via Unix-domain sockets (authentication via username)
       # We set the db password of $USER to this value which is also used by the flyway command
       FLYWAY_PASSWORD: password
+      # Extra underscore because conflicts with hardhat internally used variables.
+      HARDHAT_VERSION_: 2.6.8
+      HARDHAT_PATH_: /home/runner/hardhat
     steps:
       - uses: actions/checkout@v2
       - uses: Swatinem/rust-cache@v1
       - name: Cache flyway
+        id: cache-flyway
         uses: actions/cache@v2
         with:
-          path: flyway-${{ env.FLYWAY_VERSION }}
-          key: ${{ runner.os }}-build-flyway-${{ env.FLYWAY_VERSION }}
-      - run: "[[ -d flyway-${FLYWAY_VERSION} ]] || (curl -L https://repo1.maven.org/maven2/org/flywaydb/flyway-commandline/${FLYWAY_VERSION}/flyway-commandline-${FLYWAY_VERSION}.tar.gz -o flyway-commandline-${FLYWAY_VERSION}.tar.gz && tar -xzf flyway-commandline-${FLYWAY_VERSION}.tar.gz && rm flyway-commandline-${FLYWAY_VERSION}.tar.gz)"
-      - run: |
-          npm install hardhat
+          path: ${{ env.FLYWAY_PATH }}
+          key: ${{ runner.os }}-flyway-${{ env.FLYWAY_VERSION }}
+      - name: Install Flyway
+        if: steps.cache-flyway.outputs.cache-hit != 'true'
+        run: |
+          mkdir --parents ${{ env.FLYWAY_PATH }}
+          cd ${{ env.FLYWAY_PATH }}
+          curl -L https://repo1.maven.org/maven2/org/flywaydb/flyway-commandline/${FLYWAY_VERSION}/flyway-commandline-${FLYWAY_VERSION}.tar.gz -o out.tar.gz
+          tar --strip-components=1 -xzf out.tar.gz
+          rm out.tar.gz
+          ls -al
+      - name: Cache Hardhat
+        id: cache-hardhat
+        uses: actions/cache@v2
+        with:
+          path: ${{ env.HARDHAT_PATH_ }}/node_modules
+          key: ${{ runner.os }}-hardhat-${{ env.HARDHAT_VERSION_ }}
+      - name: Install Hardhat
+        if: steps.cache-hardhat.outputs.cache-hit != 'true'
+        run: |
+          mkdir --parents ${{ env.HARDHAT_PATH_ }}
+          cd ${{ env.HARDHAT_PATH_ }}
+          npm install hardhat@${{ env.HARDHAT_VERSION_ }}
+          ls -al
```

Comment:
Instead of installing `hardhat` in some directly, why not use the global NPM package directory and install with `npm install -g hardhat@${{ env.HARDHAT_VERSION_ }}`? The installation directory to cache would be `npm root -g` and the executable path `$(npm root -g)/node_modules/.bin/hardhat`.

Reply:
Thanks for the suggestion. Will try it out. I don't know much about node, npm so the current solution was trial and error. Didn't know about the global thing.

Reply:
I had a look. It is problematic that this directory already contains a bunch of pre installed in the CI container. We don't want to cache those but I'm not sure how to only cache hardhat. So will keep the current version for now unless someone knows how to work around that.

Reply:
Lets keep the current version then.